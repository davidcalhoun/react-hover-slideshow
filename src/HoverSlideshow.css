.container {
	touch-action: none;
	display: inline-flex;
	background-repeat: no-repeat;
	-webkit-touch-callout: none;
	-webkit-user-select: none;
}
