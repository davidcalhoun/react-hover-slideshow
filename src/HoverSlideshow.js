import React from "react";
import PropTypes from "prop-types";
import useHoverSlideshow from "./useHoverSlideshow";
import styles from "./HoverSlideshow.css";

export default function HoverSlideshow(props) {
	const { axis, images, style, ...otherProps } = props;

	let [
		{ currentImage },
		{ updateHoverSlideshow, resetHoverSlideshow }
	] = useHoverSlideshow(images, axis);

	return (
		<div
			onMouseLeave={resetHoverSlideshow}
			onMouseMove={updateHoverSlideshow}
			onTouchStart={updateHoverSlideshow}
			onTouchMove={updateHoverSlideshow}
			onTouchEnd={resetHoverSlideshow}
			className={styles.container}
			/**
			 * Swap out background-image instead of img src, the latter which is problematic
			 * with touchmove on mobile.
			 */
			style={{ backgroundImage: `url(${currentImage})`, ...style }}
			{...otherProps}
		/>
	);
}

HoverSlideshow.propTypes = {
	images: PropTypes.arrayOf(PropTypes.string)
};

HoverSlideshow.defaultProps = {
	"aria-label": "Image slideshow",
	axis: "horizontal",
	images: [],
	role: "img"
};
