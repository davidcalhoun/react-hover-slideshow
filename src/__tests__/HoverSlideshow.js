import React from "react";
import { render, fireEvent } from "@testing-library/react";

import HoverSlideshow from "../HoverSlideshow";

const imageSet1 = [
  "01.jpg",
  "02.jpg",
  "03.jpg",
  "04.jpg",
  "05.jpg",
  "06.jpg",
  "07.jpg",
  "08.jpg",
  "09.jpg",
  "10.jpg",
  "11.jpg",
  "12.jpg"
];
imageSet1.name = "imageSet1";

const imageSet2 = [
  "a.jpg",
  "b.jpg",
  "c.jpg",
  "d.jpg",
  "e.jpg",
  "f.jpg",
  "g.jpg"
];
imageSet2.name = "imageSet2";

const axes = ["horizontal", "vertical"];
const imageSets = [imageSet1, imageSet2];
const dimensions = [100, 200, 233];

dimensions.forEach(pixels => {
  describe(`${ pixels } pixels`, () => {
    imageSets.forEach(images => {
      describe(`${images.name}`, () => {
        axes.forEach(axis => {
          describe(`${axis} axis`, () => {
            const percentages = [0, 20, 50, 100];

            percentages.forEach(percent => {
              test(`${percent} percent`, async () => {
                const { container, getByLabelText } = render(
                  <HoverSlideshow axis={axis} images={images} aria-label="Test slideshow" />
                );

                const slideshow = getByLabelText("Test slideshow");

                slideshow.getBoundingClientRect = () => {
                  return {
                    width: pixels,
                    height: pixels,
                    x: 0,
                    y: 0
                  };
                };

                const mouseMoveProps =
                  axis === "horizontal"
                    ? { clientX: pixels * (percent / 100) }
                    : { clientY: pixels * (percent / 100) };

                fireEvent.mouseMove(slideshow, mouseMoveProps);

                expect(container).toMatchSnapshot();
              });
            });
          });
        });
      });
    });
  });
});
