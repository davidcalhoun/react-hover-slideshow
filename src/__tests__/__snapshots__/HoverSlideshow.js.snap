// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`100 pixels imageSet1 horizontal axis 0 percent 1`] = `
<div>
  <picture
    aria-label="Test slideshow"
    role="img"
  >
    <source
      media="(min-width: 800px)"
      srcset="01.jpg"
    />
    <img
      src="01.jpg"
    />
  </picture>
</div>
`;

exports[`100 pixels imageSet1 horizontal axis 20 percent 1`] = `
<div>
  <picture
    aria-label="Test slideshow"
    role="img"
  >
    <source
      media="(min-width: 800px)"
      srcset="03.jpg"
    />
    <img
      src="03.jpg"
    />
  </picture>
</div>
`;

exports[`100 pixels imageSet1 horizontal axis 50 percent 1`] = `
<div>
  <picture
    aria-label="Test slideshow"
    role="img"
  >
    <source
      media="(min-width: 800px)"
      srcset="06.jpg"
    />
    <img
      src="06.jpg"
    />
  </picture>
</div>
`;

exports[`100 pixels imageSet1 horizontal axis 100 percent 1`] = `
<div>
  <picture
    aria-label="Test slideshow"
    role="img"
  >
    <source
      media="(min-width: 800px)"
      srcset="12.jpg"
    />
    <img
      src="12.jpg"
    />
  </picture>
</div>
`;

exports[`100 pixels imageSet1 vertical axis 0 percent 1`] = `
<div>
  <picture
    aria-label="Test slideshow"
    role="img"
  >
    <source
      media="(min-width: 800px)"
      srcset="01.jpg"
    />
    <img
      src="01.jpg"
    />
  </picture>
</div>
`;

exports[`100 pixels imageSet1 vertical axis 20 percent 1`] = `
<div>
  <picture
    aria-label="Test slideshow"
    role="img"
  >
    <source
      media="(min-width: 800px)"
      srcset="03.jpg"
    />
    <img
      src="03.jpg"
    />
  </picture>
</div>
`;

exports[`100 pixels imageSet1 vertical axis 50 percent 1`] = `
<div>
  <picture
    aria-label="Test slideshow"
    role="img"
  >
    <source
      media="(min-width: 800px)"
      srcset="06.jpg"
    />
    <img
      src="06.jpg"
    />
  </picture>
</div>
`;

exports[`100 pixels imageSet1 vertical axis 100 percent 1`] = `
<div>
  <picture
    aria-label="Test slideshow"
    role="img"
  >
    <source
      media="(min-width: 800px)"
      srcset="12.jpg"
    />
    <img
      src="12.jpg"
    />
  </picture>
</div>
`;

exports[`100 pixels imageSet2 horizontal axis 0 percent 1`] = `
<div>
  <picture
    aria-label="Test slideshow"
    role="img"
  >
    <source
      media="(min-width: 800px)"
      srcset="a.jpg"
    />
    <img
      src="a.jpg"
    />
  </picture>
</div>
`;

exports[`100 pixels imageSet2 horizontal axis 20 percent 1`] = `
<div>
  <picture
    aria-label="Test slideshow"
    role="img"
  >
    <source
      media="(min-width: 800px)"
      srcset="b.jpg"
    />
    <img
      src="b.jpg"
    />
  </picture>
</div>
`;

exports[`100 pixels imageSet2 horizontal axis 50 percent 1`] = `
<div>
  <picture
    aria-label="Test slideshow"
    role="img"
  >
    <source
      media="(min-width: 800px)"
      srcset="d.jpg"
    />
    <img
      src="d.jpg"
    />
  </picture>
</div>
`;

exports[`100 pixels imageSet2 horizontal axis 100 percent 1`] = `
<div>
  <picture
    aria-label="Test slideshow"
    role="img"
  >
    <source
      media="(min-width: 800px)"
      srcset="g.jpg"
    />
    <img
      src="g.jpg"
    />
  </picture>
</div>
`;

exports[`100 pixels imageSet2 vertical axis 0 percent 1`] = `
<div>
  <picture
    aria-label="Test slideshow"
    role="img"
  >
    <source
      media="(min-width: 800px)"
      srcset="a.jpg"
    />
    <img
      src="a.jpg"
    />
  </picture>
</div>
`;

exports[`100 pixels imageSet2 vertical axis 20 percent 1`] = `
<div>
  <picture
    aria-label="Test slideshow"
    role="img"
  >
    <source
      media="(min-width: 800px)"
      srcset="b.jpg"
    />
    <img
      src="b.jpg"
    />
  </picture>
</div>
`;

exports[`100 pixels imageSet2 vertical axis 50 percent 1`] = `
<div>
  <picture
    aria-label="Test slideshow"
    role="img"
  >
    <source
      media="(min-width: 800px)"
      srcset="d.jpg"
    />
    <img
      src="d.jpg"
    />
  </picture>
</div>
`;

exports[`100 pixels imageSet2 vertical axis 100 percent 1`] = `
<div>
  <picture
    aria-label="Test slideshow"
    role="img"
  >
    <source
      media="(min-width: 800px)"
      srcset="g.jpg"
    />
    <img
      src="g.jpg"
    />
  </picture>
</div>
`;

exports[`200 pixels imageSet1 horizontal axis 0 percent 1`] = `
<div>
  <picture
    aria-label="Test slideshow"
    role="img"
  >
    <source
      media="(min-width: 800px)"
      srcset="01.jpg"
    />
    <img
      src="01.jpg"
    />
  </picture>
</div>
`;

exports[`200 pixels imageSet1 horizontal axis 20 percent 1`] = `
<div>
  <picture
    aria-label="Test slideshow"
    role="img"
  >
    <source
      media="(min-width: 800px)"
      srcset="03.jpg"
    />
    <img
      src="03.jpg"
    />
  </picture>
</div>
`;

exports[`200 pixels imageSet1 horizontal axis 50 percent 1`] = `
<div>
  <picture
    aria-label="Test slideshow"
    role="img"
  >
    <source
      media="(min-width: 800px)"
      srcset="06.jpg"
    />
    <img
      src="06.jpg"
    />
  </picture>
</div>
`;

exports[`200 pixels imageSet1 horizontal axis 100 percent 1`] = `
<div>
  <picture
    aria-label="Test slideshow"
    role="img"
  >
    <source
      media="(min-width: 800px)"
      srcset="12.jpg"
    />
    <img
      src="12.jpg"
    />
  </picture>
</div>
`;

exports[`200 pixels imageSet1 vertical axis 0 percent 1`] = `
<div>
  <picture
    aria-label="Test slideshow"
    role="img"
  >
    <source
      media="(min-width: 800px)"
      srcset="01.jpg"
    />
    <img
      src="01.jpg"
    />
  </picture>
</div>
`;

exports[`200 pixels imageSet1 vertical axis 20 percent 1`] = `
<div>
  <picture
    aria-label="Test slideshow"
    role="img"
  >
    <source
      media="(min-width: 800px)"
      srcset="03.jpg"
    />
    <img
      src="03.jpg"
    />
  </picture>
</div>
`;

exports[`200 pixels imageSet1 vertical axis 50 percent 1`] = `
<div>
  <picture
    aria-label="Test slideshow"
    role="img"
  >
    <source
      media="(min-width: 800px)"
      srcset="06.jpg"
    />
    <img
      src="06.jpg"
    />
  </picture>
</div>
`;

exports[`200 pixels imageSet1 vertical axis 100 percent 1`] = `
<div>
  <picture
    aria-label="Test slideshow"
    role="img"
  >
    <source
      media="(min-width: 800px)"
      srcset="12.jpg"
    />
    <img
      src="12.jpg"
    />
  </picture>
</div>
`;

exports[`200 pixels imageSet2 horizontal axis 0 percent 1`] = `
<div>
  <picture
    aria-label="Test slideshow"
    role="img"
  >
    <source
      media="(min-width: 800px)"
      srcset="a.jpg"
    />
    <img
      src="a.jpg"
    />
  </picture>
</div>
`;

exports[`200 pixels imageSet2 horizontal axis 20 percent 1`] = `
<div>
  <picture
    aria-label="Test slideshow"
    role="img"
  >
    <source
      media="(min-width: 800px)"
      srcset="b.jpg"
    />
    <img
      src="b.jpg"
    />
  </picture>
</div>
`;

exports[`200 pixels imageSet2 horizontal axis 50 percent 1`] = `
<div>
  <picture
    aria-label="Test slideshow"
    role="img"
  >
    <source
      media="(min-width: 800px)"
      srcset="d.jpg"
    />
    <img
      src="d.jpg"
    />
  </picture>
</div>
`;

exports[`200 pixels imageSet2 horizontal axis 100 percent 1`] = `
<div>
  <picture
    aria-label="Test slideshow"
    role="img"
  >
    <source
      media="(min-width: 800px)"
      srcset="g.jpg"
    />
    <img
      src="g.jpg"
    />
  </picture>
</div>
`;

exports[`200 pixels imageSet2 vertical axis 0 percent 1`] = `
<div>
  <picture
    aria-label="Test slideshow"
    role="img"
  >
    <source
      media="(min-width: 800px)"
      srcset="a.jpg"
    />
    <img
      src="a.jpg"
    />
  </picture>
</div>
`;

exports[`200 pixels imageSet2 vertical axis 20 percent 1`] = `
<div>
  <picture
    aria-label="Test slideshow"
    role="img"
  >
    <source
      media="(min-width: 800px)"
      srcset="b.jpg"
    />
    <img
      src="b.jpg"
    />
  </picture>
</div>
`;

exports[`200 pixels imageSet2 vertical axis 50 percent 1`] = `
<div>
  <picture
    aria-label="Test slideshow"
    role="img"
  >
    <source
      media="(min-width: 800px)"
      srcset="d.jpg"
    />
    <img
      src="d.jpg"
    />
  </picture>
</div>
`;

exports[`200 pixels imageSet2 vertical axis 100 percent 1`] = `
<div>
  <picture
    aria-label="Test slideshow"
    role="img"
  >
    <source
      media="(min-width: 800px)"
      srcset="g.jpg"
    />
    <img
      src="g.jpg"
    />
  </picture>
</div>
`;

exports[`233 pixels imageSet1 horizontal axis 0 percent 1`] = `
<div>
  <picture
    aria-label="Test slideshow"
    role="img"
  >
    <source
      media="(min-width: 800px)"
      srcset="01.jpg"
    />
    <img
      src="01.jpg"
    />
  </picture>
</div>
`;

exports[`233 pixels imageSet1 horizontal axis 20 percent 1`] = `
<div>
  <picture
    aria-label="Test slideshow"
    role="img"
  >
    <source
      media="(min-width: 800px)"
      srcset="03.jpg"
    />
    <img
      src="03.jpg"
    />
  </picture>
</div>
`;

exports[`233 pixels imageSet1 horizontal axis 50 percent 1`] = `
<div>
  <picture
    aria-label="Test slideshow"
    role="img"
  >
    <source
      media="(min-width: 800px)"
      srcset="06.jpg"
    />
    <img
      src="06.jpg"
    />
  </picture>
</div>
`;

exports[`233 pixels imageSet1 horizontal axis 100 percent 1`] = `
<div>
  <picture
    aria-label="Test slideshow"
    role="img"
  >
    <source
      media="(min-width: 800px)"
      srcset="12.jpg"
    />
    <img
      src="12.jpg"
    />
  </picture>
</div>
`;

exports[`233 pixels imageSet1 vertical axis 0 percent 1`] = `
<div>
  <picture
    aria-label="Test slideshow"
    role="img"
  >
    <source
      media="(min-width: 800px)"
      srcset="01.jpg"
    />
    <img
      src="01.jpg"
    />
  </picture>
</div>
`;

exports[`233 pixels imageSet1 vertical axis 20 percent 1`] = `
<div>
  <picture
    aria-label="Test slideshow"
    role="img"
  >
    <source
      media="(min-width: 800px)"
      srcset="03.jpg"
    />
    <img
      src="03.jpg"
    />
  </picture>
</div>
`;

exports[`233 pixels imageSet1 vertical axis 50 percent 1`] = `
<div>
  <picture
    aria-label="Test slideshow"
    role="img"
  >
    <source
      media="(min-width: 800px)"
      srcset="06.jpg"
    />
    <img
      src="06.jpg"
    />
  </picture>
</div>
`;

exports[`233 pixels imageSet1 vertical axis 100 percent 1`] = `
<div>
  <picture
    aria-label="Test slideshow"
    role="img"
  >
    <source
      media="(min-width: 800px)"
      srcset="12.jpg"
    />
    <img
      src="12.jpg"
    />
  </picture>
</div>
`;

exports[`233 pixels imageSet2 horizontal axis 0 percent 1`] = `
<div>
  <picture
    aria-label="Test slideshow"
    role="img"
  >
    <source
      media="(min-width: 800px)"
      srcset="a.jpg"
    />
    <img
      src="a.jpg"
    />
  </picture>
</div>
`;

exports[`233 pixels imageSet2 horizontal axis 20 percent 1`] = `
<div>
  <picture
    aria-label="Test slideshow"
    role="img"
  >
    <source
      media="(min-width: 800px)"
      srcset="b.jpg"
    />
    <img
      src="b.jpg"
    />
  </picture>
</div>
`;

exports[`233 pixels imageSet2 horizontal axis 50 percent 1`] = `
<div>
  <picture
    aria-label="Test slideshow"
    role="img"
  >
    <source
      media="(min-width: 800px)"
      srcset="d.jpg"
    />
    <img
      src="d.jpg"
    />
  </picture>
</div>
`;

exports[`233 pixels imageSet2 horizontal axis 100 percent 1`] = `
<div>
  <picture
    aria-label="Test slideshow"
    role="img"
  >
    <source
      media="(min-width: 800px)"
      srcset="g.jpg"
    />
    <img
      src="g.jpg"
    />
  </picture>
</div>
`;

exports[`233 pixels imageSet2 vertical axis 0 percent 1`] = `
<div>
  <picture
    aria-label="Test slideshow"
    role="img"
  >
    <source
      media="(min-width: 800px)"
      srcset="a.jpg"
    />
    <img
      src="a.jpg"
    />
  </picture>
</div>
`;

exports[`233 pixels imageSet2 vertical axis 20 percent 1`] = `
<div>
  <picture
    aria-label="Test slideshow"
    role="img"
  >
    <source
      media="(min-width: 800px)"
      srcset="b.jpg"
    />
    <img
      src="b.jpg"
    />
  </picture>
</div>
`;

exports[`233 pixels imageSet2 vertical axis 50 percent 1`] = `
<div>
  <picture
    aria-label="Test slideshow"
    role="img"
  >
    <source
      media="(min-width: 800px)"
      srcset="d.jpg"
    />
    <img
      src="d.jpg"
    />
  </picture>
</div>
`;

exports[`233 pixels imageSet2 vertical axis 100 percent 1`] = `
<div>
  <picture
    aria-label="Test slideshow"
    role="img"
  >
    <source
      media="(min-width: 800px)"
      srcset="g.jpg"
    />
    <img
      src="g.jpg"
    />
  </picture>
</div>
`;
